In [1]:
import pandas as pd
import numpy as np

import warnings
warnings.simplefilter(action="ignore", category=FutureWarning)

pd.set_option("display.max_rows", 10)
pd.set_option("display.max_columns", 30)

## 資料處理

In [2]:
student_survey = pd.read_csv("TEPS 2007 第四波 高中職:五專學生問卷.csv")
parent_survey = pd.read_csv("TEPS 2007 第四波 高中職:五專家長問卷.csv")

/var/folders/wd/96tcz0211cz0ppfx19qjlzrm0000gn/T/ipykernel_2875/2481485875.py:1: DtypeWarning: Columns (571,572,573,574,575,576) have mixed types. Specify dtype option on import or set low_memory=False.
  student_survey = pd.read_csv("TEPS 2007 第四波 高中職:五專學生問卷.csv")
/var/folders/wd/96tcz0211cz0ppfx19qjlzrm0000gn/T/ipykernel_2875/2481485875.py:2: DtypeWarning: Columns (16,17,18,19,20,21,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180) have mixed types. Specify dtype option on import or set low_memory=False.
  parent_survey = pd.read_csv("TEPS 2007 第四波 高中職:五專家長問卷.csv")


### 學校
| 欄位內容 | 欄位名稱 |
|---|---|
| 城鄉 | w4urban3 |
| 公私立 | w4priv |

In [3]:
schoolFactor = student_survey[["stud_id", "w4urban3", "w4priv"]]

schoolFactor

stud_id  w4urban3  w4priv
0            6         2       0
1           15         3       0
2           18         3       1
3           36         3       0
4           37         3       0
...        ...       ...     ...
13548    45865         3       1
13549    45867         2       0
13550    45869         3       1
13551    45870         3       1
13552    45872         2       0

[13553 rows x 3 columns]

### 家庭
| 欄位內容 | 欄位名稱 |
|---|---|
| 教育程度 | w4p104, 117 |
| 教養態度 | w4p216 - 229 |
| 經濟狀況 | |
| 夫妻關係 | w4p103, 114 |
| 親子關係 | w4p102, 113 |
| 性別 | w4p101 |
| 生活習慣 | w4p118 - 123 |

In [4]:
educationLevel = parent_survey[["stud_id", "w4p104", "w4p117"]]
parentingStyle = parent_survey[[f"w4p{i}" for i in range(216, 230) if i != 217 and i != 218]]  # w4p217 & w4p218 not in columns
parentRelation = parent_survey[["w4p103", "w4p114"]]
parent_childRelation = parent_survey[["w4p102", "w4p113"]]
parentGender = parent_survey[["w4p101"]]
lifeStyle = parent_survey[[f"w4p{i}" for i in range(118, 124)]]

familyFactor = pd.concat([educationLevel, parentingStyle, parentRelation, parent_childRelation, parentGender, parentingStyle], axis=1)
familyFactor

stud_id  w4p104  w4p117  w4p216  w4p219  w4p220  w4p221  w4p222  \
0            6       3       4       4       1       2       1       2   
1           15       5       3       3       1       1       2       1   
2           18       3       1       5       1       1       1       1   
3           36       3       1       4       1       1       1       1   
4           37       4       2       4       1       1       1       2   
...        ...     ...     ...     ...     ...     ...     ...     ...   
13369    45865       3       3       2       1       2       1       3   
13370    45867       4       3       4       1       1       1       1   
13371    45869       4       2       4       1       1       1       1   
13372    45870       1       1       4       1       3       1       3   
13373    45872       1       5       4       1       1       1       1   

       w4p223  w4p224  w4p225  w4p226  w4p227  w4p228  w4p229  ...  w4p102  \
0           1       2       3       3       3       2       4  ...       1   
1           2       4       1       1       2       3       2  ...       1   
2           1       1       2       2       2       3       2  ...       1   
3           1       1       4       4       4       2       4  ...       1   
4           2       1       2       2       2       2       2  ...       1   
...       ...     ...     ...     ...     ...     ...     ...  ...     ...   
13369       2       2       2       4       3       2       2  ...       1   
13370       1       1       3       3       3       1       2  ...       1   
13371       2       3       3       3       3       3       2  ...       1   
13372       1       3       3       2       3       2       4  ...       1   
13373       2       1       3       4       3       1       2  ...       1   

       w4p113  w4p101  w4p216  w4p219  w4p220  w4p221  w4p222  w4p223  w4p224  \
0           2       2       4       1       2       1       2       1       2   
1           2       1       3       1       1       2       1       2       4   
2           2       2       5       1       1       1       1       1       1   
3           1       2       4       1       1       1       1       1       1   
4           2       2       4       1       1       1       2       2       1   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
13369       1       2       2       1       2       1       3       2       2   
13370       2       1       4       1       1       1       1       1       1   
13371       1       1       4       1       1       1       1       2       3   
13372       2       2       4       1       3       1       3       1       3   
13373       1       2       4       1       1       1       1       2       1   

       w4p225  w4p226  w4p227  w4p228  w4p229  
0           3       3       3       2       4  
1           1       1       2       3       2  
2           2       2       2       3       2  
3           4       4       4       2       4  
4           2       2       2       2       2  
...       ...     ...     ...     ...     ...  
13369       2       4       3       2       2  
13370       3       3       3       1       2  
13371       3       3       3       3       2  
13372       3       2       3       2       4  
13373       3       4       3       1       2  

[13374 rows x 32 columns]

### 生活習慣
| 欄位內容 | 欄位名稱 |
|---|---|
| 生活習慣 | w4s101 - 103 |

In [5]:
life_styleFactor = student_survey[[f"w4s{i}" for i in range(101, 104)]]

life_styleFactor

w4s101  w4s102  w4s103
0           1       4       4
1           4       3       4
2           1       4       3
3           1       2       4
4           1       6       4
...       ...     ...     ...
13548       1       4       4
13549       1       2       4
13550       1       3       4
13551       1       2       3
13552       1       4       4

[13553 rows x 3 columns]

### 升學準備
| 欄位內容 | 欄位名稱 |
|---|---|
| 升學準備 | w4s104 - 107, 116 |
| 升學策略（複選題）：上高三以來，除補習之外，你採取了下列哪些策略來準備升學? | <br>w4s1281 - 1285: 策略種類</br><br>w4s1286: 無</br> |
| 升學方向（複選題）：你打算參加哪些類組的考試? | <br>w4s1301 - 1303: 類組</br><br>w4s1304 - 1305: 專、職</br><br>w4s1306: 不參加</br>

In [6]:
temp = [f"w4s{i}" for i in range(104, 108)]
temp.append("w4s116")
testPreparation = student_survey[temp]

prepareStrategy = student_survey[[f"w4s{i}" for i in range(1281, 1286)]]
testTarget = student_survey[[f"w4s{i}" for i in range(1301, 1307)]]

test_preparationFactor = pd.concat([testPreparation, prepareStrategy, testTarget], axis=1)
test_preparationFactor.insert(0, "stud_id", student_survey["stud_id"])
test_preparationFactor["w4s106"] = test_preparationFactor["w4s106"].replace(6, 0)  # 校內課輔時數
test_preparationFactor["w4s107"] = test_preparationFactor["w4s107"].replace(6, 0)  # 校外補習時數
# test_preparationFactor = test_preparationFactor[~((test_preparationFactor == 99) | (test_preparationFactor == 97)).any(axis=1)]
test_preparationFactor


stud_id  w4s104  w4s105  w4s106  w4s107  w4s116  w4s1281  w4s1282  \
0            6       6       2       1       0       2        1        1   
1           15       3       2       1       0       1        1        1   
2           18       6       1       4       0       1        0        0   
3           36       2       2       2       1       1        0        1   
4           37       1       4       3       0       6        1        0   
...        ...     ...     ...     ...     ...     ...      ...      ...   
13548    45865       3       5       5       4       1        1        1   
13549    45867       4       4       3       0       5        0        0   
13550    45869      99      99       1       0       1       99       99   
13551    45870       1       1       0       0       6        0        0   
13552    45872       2       2       2       1       1        0        0   

       w4s1283  w4s1284  w4s1285  w4s1301  w4s1302  w4s1303  w4s1304  w4s1305  \
0            0        0        0        0        0        0        1        1   
1            1        0        1        1        0        0        0        0   
2            0        0        0        0        0        0        0        1   
3            0        0        0        0        0        0        0        1   
4            0        1        1        0        0        0        0        1   
...        ...      ...      ...      ...      ...      ...      ...      ...   
13548        0        0        1        1        0        0        0        0   
13549        0        0        1        0        1        0        0        0   
13550       99       99       99       99       99       99       99       99   
13551        0        0        0        0        0        0        0        0   
13552        0        0        0        1        1        1        1        1   

       w4s1306  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
13548        0  
13549        0  
13550       99  
13551        1  
13552        0  

[13553 rows x 17 columns]

### 分析能力表現
| 欄位內容 | 欄位名稱 |
|---|---|
| 「綜合分析能力測驗」中答對的題數 | w4nright |
| 「一般分析能力測驗」中答對的題數 | w4cfree |
| 「數學分析能力測驗」中答對的題數 | w4math |

In [7]:
comprehensiveAnalysis = student_survey[["w4nright"]]
commonAnalysis = student_survey[["w4cfree"]]
mathAnalysis = student_survey[["w4math"]]

analysisFactor = pd.concat([comprehensiveAnalysis, commonAnalysis, mathAnalysis], axis=1)
analysisFactor.insert(0, "stud_id", student_survey["stud_id"])
analysisFactor = analysisFactor.dropna()
analysisFactor['w4nright'] = analysisFactor['w4nright'].replace(' ', np.nan)
analysisFactor['w4cfree'] = analysisFactor['w4cfree'].replace(' ', np.nan)
analysisFactor['w4math'] = analysisFactor['w4math'].replace(' ', np.nan)
analysisFactor['w4nright'] = analysisFactor['w4nright'].fillna(0).astype(int)
analysisFactor['w4cfree'] = analysisFactor['w4cfree'].fillna(0).astype(int)
analysisFactor['w4math'] = analysisFactor['w4math'].fillna(0).astype(int)

analysisFactor

stud_id  w4nright  w4cfree  w4math
0            6        33       13      18
1           15        30       10      14
2           18        26       11      18
3           36        28       12      13
4           37        28        7      19
...        ...       ...      ...     ...
13548    45865        35       12      22
13549    45867        40       19      23
13550    45869        13        4       6
13551    45870        23       11      13
13552    45872        24        9      12

[13553 rows x 4 columns]

### 個人因素
| 欄位內容 | 欄位名稱 |
|---|---|
| 性別 | w1s502 |
| 休閒活動種類與所花時間 | w4s117 - 119 |

In [8]:
studentGender = student_survey[["w1s502"]]
leisureTypeTime = student_survey[[f"w4s{i}" for i in range(117, 120)]]

personalFactor = pd.concat([studentGender, leisureTypeTime], axis=1)
personalFactor.insert(0, "stud_id", student_survey["stud_id"])
personalFactor = personalFactor[~((personalFactor == 99) | (personalFactor == 97)).any(axis=1)]
personalFactor

stud_id  w1s502  w4s117  w4s118  w4s119
0            6       2       6       6       3
1           15       2       1       6       3
2           18       1       4       3       2
3           36       1       1       1       3
4           37       2       6       6       3
...        ...     ...     ...     ...     ...
13548    45865       2       6       6       3
13549    45867       1       5       3       4
13550    45869       2       1       6       3
13551    45870       1       3       1       5
13552    45872       1       5       5       3

[13417 rows x 5 columns]

## 迴歸分析

In [9]:
import matplotlib.pyplot as plt
from matplotlib.pyplot import figure

import statsmodels.formula.api as sm
import statsmodels.api as sm

import sklearn
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics

### 學校因素
#### 公私立 vs. 校內課輔參與時數

In [10]:
reg1_1 = pd.concat([test_preparationFactor["w4s106"], test_preparationFactor["w4s107"], schoolFactor["w4priv"]], axis=1)
reg1_1 = reg1_1[~((reg1_1 == 99) | (reg1_1 == 97)).any(axis=1)]

X = reg1_1.drop("w4s106", axis=1)
y = reg1_1["w4s106"]

X = sm.add_constant(X)  # 加入截距讓迴歸更準確

model1_1 = sm.OLS(y, X)
result = model1_1.fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                 w4s106   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.006
Method:                 Least Squares   F-statistic:                     43.82
Date:                Wed, 31 May 2023   Prob (F-statistic):           1.07e-19
Time:                        00:57:10   Log-Likelihood:                -23053.
No. Observations:               13500   AIC:                         4.611e+04
Df Residuals:                   13497   BIC:                         4.613e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5785      0.018     86.813      0.0

#### 公私立 vs. 校外補習參與時數

In [11]:
reg1_2 = pd.concat([test_preparationFactor["w4s106"], test_preparationFactor["w4s107"], schoolFactor["w4priv"]], axis=1)
reg1_2 = reg1_2.loc[:,~reg1_2.columns.duplicated()]
reg1_2 = reg1_2[~((reg1_2 == 99) | (reg1_2 == 97)).any(axis=1)]

X = reg1_2.drop("w4s107", axis=1)
y = reg1_2["w4s107"]

X = sm.add_constant(X)  # 加入截距讓迴歸更準確

model1_2 = sm.OLS(y, X)
result = model1_2.fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                 w4s107   R-squared:                       0.073
Model:                            OLS   Adj. R-squared:                  0.073
Method:                 Least Squares   F-statistic:                     535.1
Date:                Wed, 31 May 2023   Prob (F-statistic):          2.25e-224
Time:                        00:57:10   Log-Likelihood:                -24385.
No. Observations:               13500   AIC:                         4.878e+04
Df Residuals:                   13497   BIC:                         4.880e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.2933      0.022     57.628      0.0

#### 類組 & 公私立 vs. 補習時數

### 家庭因素

#### 家長教育程度 & 升學準備策略、時間 vs. 類組選擇
- 包含家長教育程度 => Pseudo R-squ: 0.1491
- 不包含家長教育程度 => Pseudo R-squ: 0.1389

In [12]:
# 包含家長教育程度 => Pseudo R-squ.: 0.1491
reg2_1 = pd.merge(test_preparationFactor, educationLevel[["stud_id", "w4p104"]], on="stud_id")

conditions = [
    reg2_1["w4s1301"] == 1,
    reg2_1["w4s1302"] == 1,
    reg2_1["w4s1303"] == 1,
    reg2_1["w4s1304"] == 1,
    reg2_1["w4s1305"] == 1,
    reg2_1["w4s1306"] == 1
]
choices = [1, 2, 3, 4, 5, 0]
reg2_1["group"] = np.select(conditions, choices, default=None)

reg2_1.drop([f"w4s{i}" for i in range(1301, 1307)], axis=1, inplace=True)
reg2_1 = reg2_1[~((reg2_1 == 99) | (reg2_1 == 97)).any(axis=1)]
reg2_1 = reg2_1.dropna(subset=["group"])

X = reg2_1.drop(["group"], axis=1)
y = reg2_1["group"]
y = y.astype('int')

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2023, stratify=y)

model2_1 = LogisticRegression(random_state=0, multi_class="multinomial", penalty="none", solver="newton-cg").fit(X_train, y_train)
preds = model2_1.predict(X_test)

params = model2_1.get_params()
# print(params)

#Print model parameters
# print('Intercept: \n', model2_1.intercept_)
# print('Coefficients: \n', model2_1.coef_)

#Calculate odds ratio estimates
np.exp(model2_1.coef_)

#Use statsmodels to assess variables
logit_model=sm.MNLogit(y_train,sm.add_constant(X_train))
logit_model
result=logit_model.fit()
stats1=result.summary()
stats2=result.summary2()
print(stats1)
print(stats2)

/opt/anaconda3/lib/python3.9/site-packages/scipy/optimize/linesearch.py:478: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
/opt/anaconda3/lib/python3.9/site-packages/scipy/optimize/linesearch.py:327: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/optimize.py:210: ConvergenceWarning: newton-cg failed to converge. Increase the number of iterations.
  warnings.warn(


Optimization terminated successfully.
         Current function value: 1.252981
         Iterations 8
                          MNLogit Regression Results                          
Dep. Variable:                  group   No. Observations:                 9264
Model:                        MNLogit   Df Residuals:                     9199
Method:                           MLE   Df Model:                           60
Date:                Wed, 31 May 2023   Pseudo R-squ.:                  0.1491
Time:                        00:57:15   Log-Likelihood:                -11608.
converged:                       True   LL-Null:                       -13642.
Covariance Type:            nonrobust   LLR p-value:                     0.000
   group=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.4198      0.317      4.479      0.000       0.798       2.041
stud_id    -1.798e-06   5.21e

In [13]:
# 不包含家長教育程度 => Pseudo R-squ.: 0.1389
reg2_1 = pd.merge(test_preparationFactor, educationLevel[["stud_id"]], on="stud_id")

conditions = [
    reg2_1["w4s1301"] == 1,
    reg2_1["w4s1302"] == 1,
    reg2_1["w4s1303"] == 1,
    reg2_1["w4s1304"] == 1,
    reg2_1["w4s1305"] == 1,
    reg2_1["w4s1306"] == 1
]
choices = [1, 2, 3, 4, 5, 0]
reg2_1["group"] = np.select(conditions, choices, default=None)

reg2_1.drop([f"w4s{i}" for i in range(1301, 1307)], axis=1, inplace=True)
reg2_1 = reg2_1[~((reg2_1 == 99) | (reg2_1 == 97)).any(axis=1)]
reg2_1 = reg2_1.dropna(subset=["group"])

X = reg2_1.drop(["group"], axis=1)
y = reg2_1["group"]
y = y.astype('int')

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2023, stratify=y)

model2_1 = LogisticRegression(random_state=0, multi_class="multinomial", penalty="none", solver="newton-cg").fit(X_train, y_train)
preds = model2_1.predict(X_test)

params = model2_1.get_params()
# print(params)

#Print model parameters
# print('Intercept: \n', model2_1.intercept_)
# print('Coefficients: \n', model2_1.coef_)

#Calculate odds ratio estimates
np.exp(model2_1.coef_)

#Use statsmodels to assess variables
logit_model=sm.MNLogit(y_train,sm.add_constant(X_train))
logit_model
result=logit_model.fit()
stats1=result.summary()
stats2=result.summary2()
print(stats1)
print(stats2)

/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/optimize.py:210: ConvergenceWarning: newton-cg failed to converge. Increase the number of iterations.
  warnings.warn(


Optimization terminated successfully.
         Current function value: 1.268237
         Iterations 8
                          MNLogit Regression Results                          
Dep. Variable:                  group   No. Observations:                 9269
Model:                        MNLogit   Df Residuals:                     9209
Method:                           MLE   Df Model:                           55
Date:                Wed, 31 May 2023   Pseudo R-squ.:                  0.1389
Time:                        00:57:17   Log-Likelihood:                -11755.
converged:                       True   LL-Null:                       -13651.
Covariance Type:            nonrobust   LLR p-value:                     0.000
   group=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.6530      0.295      5.605      0.000       1.075       2.231
stud_id    -2.379e-06    5.2e

### 分析能力表現
- 含全部三種分析能力 => Pseudo R-squ: 0.08488
- 只含「綜合分析能力」 => Pseudo R-squ: 0.09783
- 只含「一般分析能力」 => Pseudo R-squ: 0.07659
- 只含「數學分析能力」 => Pseudo R-squ: 0.08752

#### 分析能力表現 & 性別 vs. 類組選擇

In [14]:
reg4_1 = pd.merge(test_preparationFactor[["stud_id", "w4s1301", "w4s1302", "w4s1303", "w4s1304", "w4s1305", "w4s1306"]], personalFactor[["stud_id", "w1s502"]], on="stud_id")
reg4_1 = pd.merge(analysisFactor, reg4_1, on="stud_id")
reg4_1["w1s502"] = reg4_1["w1s502"].replace(2, 0)
reg4_1 = reg4_1.rename(columns={"w1s502": "isMale"})

conditions = [
    reg4_1["w4s1301"] == 1,
    reg4_1["w4s1302"] == 1,
    reg4_1["w4s1303"] == 1,
    reg4_1["w4s1304"] == 1,
    reg4_1["w4s1305"] == 1,
    reg4_1["w4s1306"] == 1
]
choices = [1, 2, 3, 4, 5, 0]
reg4_1["group"] = np.select(conditions, choices, default=None)

reg4_1.drop([f"w4s{i}" for i in range(1301, 1307)], axis=1, inplace=True)
reg4_1 = reg4_1.dropna(subset=["group"])

X = reg4_1.drop(["group"], axis=1)
y = reg4_1["group"]
y = y.astype('int')

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2023, stratify=y)

model4_1 = LogisticRegression(random_state=0, multi_class="multinomial", penalty="none", solver="newton-cg").fit(X_train, y_train)
preds = model4_1.predict(X_test)

params = model4_1.get_params()
# print(params)

#Print model parameters
# print('Intercept: \n', model2_1.intercept_)
# print('Coefficients: \n', model2_1.coef_)

#Calculate odds ratio estimates
np.exp(model4_1.coef_)

#Use statsmodels to assess variables
logit_model=sm.MNLogit(y_train,sm.add_constant(X_train))
logit_model
result=logit_model.fit()
stats1=result.summary()
stats2=result.summary2()
print(stats1)
print()
print(stats2)

/opt/anaconda3/lib/python3.9/site-packages/scipy/optimize/linesearch.py:478: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
/opt/anaconda3/lib/python3.9/site-packages/scipy/optimize/linesearch.py:327: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/optimize.py:210: ConvergenceWarning: newton-cg failed to converge. Increase the number of iterations.
  warnings.warn(


Optimization terminated successfully.
         Current function value: 1.324756
         Iterations 7
                          MNLogit Regression Results                          
Dep. Variable:                  group   No. Observations:                 9519
Model:                        MNLogit   Df Residuals:                     9489
Method:                           MLE   Df Model:                           25
Date:                Wed, 31 May 2023   Pseudo R-squ.:                  0.1038
Time:                        00:57:19   Log-Likelihood:                -12610.
converged:                       True   LL-Null:                       -14072.
Covariance Type:            nonrobust   LLR p-value:                     0.000
   group=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.2198      0.218     -1.009      0.313      -0.647       0.207
stud_id     8.969e-08   4.89e

#### 分析能力類型 vs. 類組選擇

##### 三種分析能力：綜合、一般、數學

In [20]:
# 三種分析能力 => Pseudo R-squ.: 0.08488
reg4_1 = pd.merge(test_preparationFactor[["stud_id", "w4s1301", "w4s1302", "w4s1303", "w4s1304", "w4s1305", "w4s1306"]], analysisFactor, on="stud_id")

conditions = [
    reg4_1["w4s1301"] == 1,
    reg4_1["w4s1302"] == 1,
    reg4_1["w4s1303"] == 1,
    reg4_1["w4s1304"] == 1,
    reg4_1["w4s1305"] == 1,
    reg4_1["w4s1306"] == 1
]
choices = [1, 2, 3, 4, 5, 0]
reg4_1["group"] = np.select(conditions, choices, default=None)

reg4_1.drop([f"w4s{i}" for i in range(1301, 1307)], axis=1, inplace=True)
reg4_1 = reg4_1.dropna(subset=["group"])

X = reg4_1.drop(["group", "stud_id"], axis=1)
y = reg4_1["group"]
y = y.astype('int')

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2023, stratify=y)

model4_1 = LogisticRegression(random_state=0, multi_class="multinomial", penalty="none", solver="newton-cg").fit(X_train, y_train)
preds = model4_1.predict(X_test)

params = model4_1.get_params()
# print(params)

#Print model parameters
# print('Intercept: \n', model2_1.intercept_)
# print('Coefficients: \n', model2_1.coef_)

#Calculate odds ratio estimates
np.exp(model4_1.coef_)

#Use statsmodels to assess variables
logit_model=sm.MNLogit(y_train,sm.add_constant(X_train))
logit_model
result=logit_model.fit()
stats1=result.summary()
stats2=result.summary2()
print(stats1)
print()
print(stats2)

Optimization terminated successfully.
         Current function value: 1.353995
         Iterations 7
                          MNLogit Regression Results                          
Dep. Variable:                  group   No. Observations:                 9552
Model:                        MNLogit   Df Residuals:                     9532
Method:                           MLE   Df Model:                           15
Date:                Wed, 31 May 2023   Pseudo R-squ.:                 0.08413
Time:                        00:57:34   Log-Likelihood:                -12933.
converged:                       True   LL-Null:                       -14121.
Covariance Type:            nonrobust   LLR p-value:                     0.000
   group=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.6256      0.130     -4.807      0.000      -0.881      -0.371
w4nright       0.1905      0.

##### 只含「綜合分析能力」 => Pseudo R-squ.: 0.07556

In [16]:
# 含「性別」＆「綜合分析能力」 => Pseudo R-squ.: 0.09783
# 只含「綜合分析能力」 => Pseudo R-squ.:0.07556
reg4_1 = pd.merge(test_preparationFactor[["stud_id", "w4s1301", "w4s1302", "w4s1303", "w4s1304", "w4s1305", "w4s1306"]], personalFactor[["stud_id", "w1s502"]], on="stud_id")
reg4_1 = pd.merge(analysisFactor[["stud_id", "w4nright"]], reg4_1, on="stud_id")
reg4_1["w1s502"] = reg4_1["w1s502"].replace(2, 0)
reg4_1 = reg4_1.rename(columns={"w1s502": "isMale"})

conditions = [
    reg4_1["w4s1301"] == 1,
    reg4_1["w4s1302"] == 1,
    reg4_1["w4s1303"] == 1,
    reg4_1["w4s1304"] == 1,
    reg4_1["w4s1305"] == 1,
    reg4_1["w4s1306"] == 1
]
choices = [1, 2, 3, 4, 5, 0]
reg4_1["group"] = np.select(conditions, choices, default=None)

reg4_1.drop(["stud_id", "isMale"], axis=1, inplace=True)
reg4_1.drop([f"w4s{i}" for i in range(1301, 1307)], axis=1, inplace=True)
reg4_1 = reg4_1.dropna(subset=["group"])

X = reg4_1.drop(["group"], axis=1)
y = reg4_1["group"]
y = y.astype('int')

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2023, stratify=y)

model4_1 = LogisticRegression(random_state=0, multi_class="multinomial", penalty="none", solver="newton-cg").fit(X_train, y_train)
preds = model4_1.predict(X_test)

params = model4_1.get_params()
# print(params)

#Print model parameters
# print('Intercept: \n', model2_1.intercept_)
# print('Coefficients: \n', model2_1.coef_)

#Calculate odds ratio estimates
np.exp(model4_1.coef_)

#Use statsmodels to assess variables
logit_model=sm.MNLogit(y_train,sm.add_constant(X_train))
logit_model
result=logit_model.fit()
stats1=result.summary()
stats2=result.summary2()
print(stats1)
print()
print(stats2)

Optimization terminated successfully.
         Current function value: 1.366568
         Iterations 7
                          MNLogit Regression Results                          
Dep. Variable:                  group   No. Observations:                 9519
Model:                        MNLogit   Df Residuals:                     9509
Method:                           MLE   Df Model:                            5
Date:                Wed, 31 May 2023   Pseudo R-squ.:                 0.07556
Time:                        00:57:22   Log-Likelihood:                -13008.
converged:                       True   LL-Null:                       -14072.
Covariance Type:            nonrobust   LLR p-value:                     0.000
   group=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.5832      0.129     -4.508      0.000      -0.837      -0.330
w4nright       0.1177      0.

##### 只含「一般分析能力」 => Pseudo R-squ.: 0.07659

In [17]:
# 含「性別」＆「一般分析能力」 => Pseudo R-squ.: 0.07659
# 只含「一般分析能力」 => Pseudo R-squ.: 0.05385
reg4_1 = pd.merge(test_preparationFactor[["stud_id", "w4s1301", "w4s1302", "w4s1303", "w4s1304", "w4s1305", "w4s1306"]], personalFactor[["stud_id", "w1s502"]], on="stud_id")
reg4_1 = pd.merge(analysisFactor[["stud_id", "w4cfree"]], reg4_1, on="stud_id")
reg4_1["w1s502"] = reg4_1["w1s502"].replace(2, 0)
reg4_1 = reg4_1.rename(columns={"w1s502": "isMale"})

conditions = [
    reg4_1["w4s1301"] == 1,
    reg4_1["w4s1302"] == 1,
    reg4_1["w4s1303"] == 1,
    reg4_1["w4s1304"] == 1,
    reg4_1["w4s1305"] == 1,
    reg4_1["w4s1306"] == 1
]
choices = [1, 2, 3, 4, 5, 0]
reg4_1["group"] = np.select(conditions, choices, default=None)

reg4_1.drop([f"w4s{i}" for i in range(1301, 1307)], axis=1, inplace=True)
reg4_1 = reg4_1.dropna(subset=["group"])
reg4_1.drop(["stud_id", "isMale"], axis=1, inplace=True)

X = reg4_1.drop(["group"], axis=1)
y = reg4_1["group"]
y = y.astype('int')

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2023, stratify=y)

model4_1 = LogisticRegression(random_state=0, multi_class="multinomial", penalty="none", solver="newton-cg").fit(X_train, y_train)
preds = model4_1.predict(X_test)

params = model4_1.get_params()
# print(params)

#Print model parameters
# print('Intercept: \n', model2_1.intercept_)
# print('Coefficients: \n', model2_1.coef_)

#Calculate odds ratio estimates
np.exp(model4_1.coef_)

#Use statsmodels to assess variables
logit_model=sm.MNLogit(y_train,sm.add_constant(X_train))
logit_model
result=logit_model.fit()
stats1=result.summary()
stats2=result.summary2()
print(stats1)
print()
print(stats2)

Optimization terminated successfully.
         Current function value: 1.398653
         Iterations 7
                          MNLogit Regression Results                          
Dep. Variable:                  group   No. Observations:                 9519
Model:                        MNLogit   Df Residuals:                     9509
Method:                           MLE   Df Model:                            5
Date:                Wed, 31 May 2023   Pseudo R-squ.:                 0.05385
Time:                        00:57:23   Log-Likelihood:                -13314.
converged:                       True   LL-Null:                       -14072.
Covariance Type:            nonrobust   LLR p-value:                     0.000
   group=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0496      0.121      0.409      0.682      -0.188       0.287
w4cfree        0.2372      0.

##### 只含「數學分析能力」 => Pseudo R-squ.: 0.08752

In [18]:
# 含「性別」＆「數學分析能力」 => Pseudo R-squ.: 0.08752
# 只含「數學分析能力」 => Pseudo R-squ.: 0.06569
reg4_1 = pd.merge(test_preparationFactor[["stud_id", "w4s1301", "w4s1302", "w4s1303", "w4s1304", "w4s1305", "w4s1306"]], personalFactor[["stud_id", "w1s502"]], on="stud_id")
reg4_1 = pd.merge(analysisFactor[["stud_id", "w4math"]], reg4_1, on="stud_id")
reg4_1["w1s502"] = reg4_1["w1s502"].replace(2, 0)
reg4_1 = reg4_1.rename(columns={"w1s502": "isMale"})

conditions = [
    reg4_1["w4s1301"] == 1,
    reg4_1["w4s1302"] == 1,
    reg4_1["w4s1303"] == 1,
    reg4_1["w4s1304"] == 1,
    reg4_1["w4s1305"] == 1,
    reg4_1["w4s1306"] == 1
]
choices = [1, 2, 3, 4, 5, 0]
reg4_1["group"] = np.select(conditions, choices, default=None)

reg4_1.drop([f"w4s{i}" for i in range(1301, 1307)], axis=1, inplace=True)
reg4_1 = reg4_1.dropna(subset=["group"])
reg4_1.drop(["stud_id", "isMale"], axis=1, inplace=True)

X = reg4_1.drop(["group"], axis=1)
y = reg4_1["group"]
y = y.astype('int')

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2023, stratify=y)

model4_1 = LogisticRegression(random_state=0, multi_class="multinomial", penalty="none", solver="newton-cg").fit(X_train, y_train)
preds = model4_1.predict(X_test)

params = model4_1.get_params()
# print(params)

#Print model parameters
# print('Intercept: \n', model2_1.intercept_)
# print('Coefficients: \n', model2_1.coef_)

#Calculate odds ratio estimates
np.exp(model4_1.coef_)

#Use statsmodels to assess variables
logit_model=sm.MNLogit(y_train,sm.add_constant(X_train))
logit_model
result=logit_model.fit()
stats1=result.summary()
stats2=result.summary2()
print(stats1)
print()
print(stats2)

Optimization terminated successfully.
         Current function value: 1.381156
         Iterations 7
                          MNLogit Regression Results                          
Dep. Variable:                  group   No. Observations:                 9519
Model:                        MNLogit   Df Residuals:                     9509
Method:                           MLE   Df Model:                            5
Date:                Wed, 31 May 2023   Pseudo R-squ.:                 0.06569
Time:                        00:57:24   Log-Likelihood:                -13147.
converged:                       True   LL-Null:                       -14072.
Covariance Type:            nonrobust   LLR p-value:                     0.000
   group=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0237      0.118      0.201      0.841      -0.208       0.255
w4math         0.1699      0.

### 個人因素

#### 性別 vs. 類組選擇

- 性別 => Pseudo R-squ.: 0.02316
- 分析能力表現 & 性別 => Pseudo R-squ.: 0.1038

In [19]:
reg3_1 = pd.merge(test_preparationFactor[["stud_id", "w4s1301", "w4s1302", "w4s1303", "w4s1304", "w4s1305", "w4s1306"]], personalFactor[["stud_id", "w1s502"]], on="stud_id")
reg3_1["w1s502"] = reg3_1["w1s502"].replace(2, 0)
reg3_1 = reg3_1.rename(columns={"w1s502": "isMale"})

conditions = [
    reg3_1["w4s1301"] == 1,
    reg3_1["w4s1302"] == 1,
    reg3_1["w4s1303"] == 1,
    reg3_1["w4s1304"] == 1,
    reg3_1["w4s1305"] == 1,
    reg3_1["w4s1306"] == 1
]
choices = [1, 2, 3, 4, 5, 0]
reg3_1["group"] = np.select(conditions, choices, default=None)

reg3_1.drop([f"w4s{i}" for i in range(1301, 1307)], axis=1, inplace=True)
reg3_1 = reg3_1[~((reg3_1 == 99) | (reg3_1 == 97)).any(axis=1)]
reg3_1 = reg3_1.dropna(subset=["group"])

reg3_1
X = reg3_1.drop(["group"], axis=1)
y = reg3_1["group"]
y = y.astype('int')

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2023, stratify=y)

model3_1 = LogisticRegression(random_state=0, multi_class="multinomial", penalty="none", solver="newton-cg").fit(X_train, y_train)
preds = model3_1.predict(X_test)

params = model3_1.get_params()
# print(params)

#Print model parameters
# print('Intercept: \n', model2_1.intercept_)
# print('Coefficients: \n', model2_1.coef_)

#Calculate odds ratio estimates
np.exp(model3_1.coef_)

#Use statsmodels to assess variables
logit_model=sm.MNLogit(y_train,sm.add_constant(X_train))
logit_model
result=logit_model.fit()
stats1=result.summary()
stats2=result.summary2()
print(stats1)
print(stats2)

/opt/anaconda3/lib/python3.9/site-packages/scipy/optimize/linesearch.py:478: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
/opt/anaconda3/lib/python3.9/site-packages/scipy/optimize/linesearch.py:327: LineSearchWarning: The line search algorithm did not converge
  warn('The line search algorithm did not converge', LineSearchWarning)
/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/optimize.py:210: ConvergenceWarning: newton-cg failed to converge. Increase the number of iterations.
  warnings.warn(


Optimization terminated successfully.
         Current function value: 1.444022
         Iterations 6
                          MNLogit Regression Results                          
Dep. Variable:                  group   No. Observations:                 9519
Model:                        MNLogit   Df Residuals:                     9504
Method:                           MLE   Df Model:                           10
Date:                Wed, 31 May 2023   Pseudo R-squ.:                 0.02316
Time:                        00:57:25   Log-Likelihood:                -13746.
converged:                       True   LL-Null:                       -14072.
Covariance Type:            nonrobust   LLR p-value:                1.372e-133
   group=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9720      0.172     17.276      0.000       2.635       3.309
stud_id    -5.515e-06   4.61e